crates/interpreter/src/util.rs

```diff
@@ -12,10 +12,124 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use crate::error::*;
+
 // TODO(dev/fast-interp): Add debug asserts when `off` is positive and negative, and `toctou`
 // support.
 pub fn offset_front<T>(cur: &[T], off: isize) -> &[T] {
     unsafe {
         core::slice::from_raw_parts(cur.as_ptr().offset(off), (cur.len() as isize - off) as usize)
     }
 }
+
+#[allow(dead_code)]
+pub fn offset_front_check<'a, T>(beg: &'a [T], cur: &'a [T], off: isize) -> Result<&'a [T], Error> {
+    if beg.is_empty() || cur.is_empty() {
+        #[cfg(feature = "debug")]
+        eprintln!("One or both slices are empty.");
+        return Err(invalid());
+    }
+    if beg.last().unwrap() as *const T != cur.last().unwrap() as *const T {
+        #[cfg(feature = "debug")]
+        eprintln!("The last elements of beg and cur do not match.");
+        return Err(invalid());
+    }
+    let mut cur_offset = beg.len().checked_sub(cur.len()).ok_or_else(|| {
+        #[cfg(feature = "debug")]
+        eprintln!("Cur length {} exceeds beg length {}.", cur.len(), beg.len());
+        invalid()
+    })?;
+    if off > 0 {
+        cur_offset = cur_offset.checked_add(off as usize).ok_or_else(|| {
+            #[cfg(feature = "debug")]
+            eprintln!("Addition overflow");
+            invalid()
+        })?;
+        if cur_offset < cur.len() {
+            #[cfg(feature = "debug")]
+            eprintln!("The offset overflows the slice length.");
+            return Err(invalid());
+        }
+    } else {
+        cur_offset = cur_offset.checked_sub(off as usize).ok_or_else(|| {
+            #[cfg(feature = "debug")]
+            eprintln!("Subtraction overflow");
+            invalid()
+        })?;
+        if cur_offset >= cur.len() {
+            #[cfg(feature = "debug")]
+            eprintln!("The offset overflows below 0.");
+            return Err(invalid());
+        }
+    }
+    Ok(&beg[cur_offset ..])
+}
+
+#[allow(dead_code)]
+#[allow(unused_variables)]
+pub fn offset_front_use<'a, T>(beg: &'a [T], cur: &'a [T], off: isize) -> &'a [T] {
+    #[cfg(not(feature = "toctou"))]
+    return offset_front_unsafe(beg, off);
+    #[cfg(feature = "toctou")]
+    offset_front_toctou(beg, cur, off)
+}
+
+#[allow(dead_code)]
+fn offset_front_toctou<'a, T>(beg: &'a [T], cur: &'a [T], off: isize) -> &'a [T] {
```

Comment:
This could theoretically just be `offset_front_check(beg, cur, off).unwrap()`. But I'm pushing another relatively simple implementation which doesn't check the end.